functions/src/db/addTokenToUser.ts

```diff
@@ -0,0 +1,33 @@
+import { UID } from '../types/uid'
+import { JWTToken } from '../types/JWTToken'
+import { db, serverTimestamp } from '../firebase/firebase'
+
+export const addTokenToUser = async (
+    userId: UID,
+    token: JWTToken
+): Promise<{ success: boolean }> => {
```

Comment:
why returning an object instead of directly a `boolean` here ?

Reply:
it makes things more clear about what's expected for the return.

Reply:
not sure I agree on this one. Why not to:
- raise an `Error` when the operation fails
- do not return anything in case the operation succeeds

Reply:
my thinking is that throwing error are not convenient to manage with async/await, and having a clear idea of the result just simplify reading the code later on.

Reply:
I see, but doesn't actually `async/await` allow to write a code like it was synchronous ? So what is the difference with `try/catch` error handling in asynchronous compared to synchronous code ?

My thinking is that it is easier to read error cases as error raised by the function, instead of returning a response no matter error/success case, than having to interpret the read and interpret the result

Reply:
Two more arguments: 
1. You want to authentificate a user. They are two possible outputs: the user is authentificated, or he is not. Both are expected state, managed by the code. So a method like `authentificateUser(): boolean` could, if using Error, never return false because it will throw an Error if the auth did not work. And as JS/IDE are nore powerful enough, how would you check for the type of value returned by the function? Is a boolean an indication that the user is logged in or not? Having a strong typed return value make things more clear. And throwing an error mean than this was not expected, but it was, it is just a rejection, not an exception. 

2. How would you finely manage error using multiple async/await within the same method? So having a try/catch for each await is possible, but it's very verbosed, and in my opinion make code harder to read.

But on the other side, we could do something like this: https://www.smashingmagazine.com/2020/08/error-handling-nodejs-error-classes/#error-classes

Reply:
My point of view on the arguments:
1. Indeed there's still the need for functions that return boolean values, and that's a good example. There are two possible states for a user in your example, none of which is an error, and the function correctly express that: `isAuthenticated` would either return `true` or `false`, according to the user's state. In the case I commented instead, it seems you're not verifying something but trying to perform an action: `addTokenToUser`. This action can either succeed or fail. If you try to perform an action, you need to know the action succeeded, in which case the normal flow would continue (return a value if some valuable information is returned by the function). You also need to know if the execution failed, in which case you can throw a specific error, and handle it in a catch block.

2. I would indeed use error-classes, as I proposed in the other comment. That allows to:
    - have a single error handling block for a function (that calls multiple async functions)
    - specify what kind of error occurred in a reusable and clear way
    - distinguish different kind of errors
    - have a single error exit point for the function